sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeOuterJoin.scala

```diff
@@ -271,3 +297,189 @@ private class RightOuterIterator(
+  advancedRight()
+
+  // --- Private methods --------------------------------------------------------------------------
+
+  /**
+   * Advance the left iterator and compute the new row's join key.
+   * @return true if the left iterator returned a row and false otherwise.
+   */
+  private def advancedLeft(): Boolean = {
+    if (leftIter.advanceNext()) {
+      leftRow = leftIter.getRow
+      leftRowKey = leftKeyGenerator(leftRow)
+      numLeftRows += 1
+      true
+    } else {
+      leftRow = null
+      leftRowKey = null
+      false
+    }
+  }
+
+  /**
+   * Advance the right iterator and compute the new row's join key.
+   * @return true if the right iterator returned a row and false otherwise.
+   */
+  private def advancedRight(): Boolean = {
+    if (rightIter.advanceNext()) {
+      rightRow = rightIter.getRow
+      rightRowKey = rightKeyGenerator(rightRow)
+      numRightRows += 1
+      true
+    } else {
+      rightRow = null
+      rightRowKey = null
+      false
+    }
+  }
+
+  /**
+   * Consume rows from both iterators until their keys are different from the provided matchingKey.
+   */
+  private def findMatchingRows(matchingKey: InternalRow): Unit = {
+    leftMatches.clear()
+    rightMatches.clear()
+    leftIndex = 0
+    rightIndex = 0
+
+    while (leftRowKey != null && keyOrdering.compare(leftRowKey, matchingKey) == 0) {
+      leftMatches += leftRow.copy()
+      advancedLeft()
+    }
+    while (rightRowKey != null && keyOrdering.compare(rightRowKey, matchingKey) == 0) {
+      rightMatches += rightRow.copy()
+      advancedRight()
+    }
+
+    if (leftMatches.size <= leftMatched.capacity) {
+      leftMatched.clear()
+    } else {
+      leftMatched = new BitSet(leftMatches.size)
+    }
+    if (rightMatches.size <= rightMatched.capacity) {
+      rightMatched.clear()
+    } else {
+      rightMatched = new BitSet(rightMatches.size)
+    }
+  }
+
+  /**
+   * Scan for next matching rows in buffers of both sides.
+   * @return true if next row(s) are found and false otherwise.
+   */
+  private def scanNextInBuffered(): Boolean = {
+    while (leftIndex < leftMatches.size) {
+      while (rightIndex < rightMatches.size
+        && !boundCondition(joinedRow(leftMatches(leftIndex), rightMatches(rightIndex)))) {
+        rightIndex += 1
+      }
+      if (rightIndex < rightMatches.size) {
+        // matched with condition
+        leftMatched.set(leftIndex)
+        rightMatched.set(rightIndex)
+        rightIndex += 1
+        return true
+      }
+      leftIndex += 1
+      rightIndex = 0
+      if (!leftMatched.get(leftIndex - 1)) {
+        joinedRow(leftMatches(leftIndex - 1), rightNullRow)
+        return true
+      }
+    }
+
+    while (rightIndex < rightMatches.size && rightMatched.get(rightIndex)) {
+      rightIndex += 1
+    }
+    if (rightIndex < rightMatches.size) {
+      joinedRow(leftNullRow, rightMatches(rightIndex))
+      rightIndex += 1
+      true
+    } else {
+      false
+    }
```

Comment:
The assert is not true, because it can start with leftIndex = leftMatches.size and 0< rightIndex <= rightMatches.size